### PR TITLE
Adds an optional application field to the job description

### PR DESCRIPTION
**Pull request by dposada**
Resolves issue https://github.com/twosigma/Cook/issues/192

### CODE REVIEW

File: scheduler/src/cook/mesos/api.clj [hunk 1]
```diff
@@ -170,13 +170,28 @@
...
+   :version (s/constrained s/Str non-empty-max-128-characters-and-alphanum?)})
+
 (def Job
   "A schema for a job"
   {:uuid s/Uuid
    :command s/Str
-   ;; Make sure the job name is a valid string which can only contain '.', '_', '-' or any word characters and has
-   ;; length at most 128.
-   :name (s/both s/Str (s/pred #(re-matches #"[\.a-zA-Z0-9_-]{0,128}" %) 'under-128-characters-and-alphanum?))
+   :name (s/both s/Str (s/pred max-128-characters-and-alphanum? 'max-128-characters-and-alphanum?))
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: I think it makes sense for name to be non-empty (I know it wasn't)

File: scheduler/src/cook/mesos/api.clj [hunk 2]
```diff
@@ -349,12 +365,21 @@
                 (mk-docker-ports docker-id port-mappings))])
       {})))
 
+(defn- build-application
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: If you make application a component (), then you can set the value of the key :job/application to the following map

```clojure
{:application/name name
 :application/version version}
```

Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: For example, see the resource list in make-job-txn

File: scheduler/src/cook/mesos/api.clj [hunk 3]
```diff
@@ -650,79 +677,81 @@
...
+                              base (if reason
+                                     (assoc base
+                                       :reason_code (:reason/code reason)
+                                       :reason_string (:reason/string reason))
+                                     base)]
+                          base))
+                      (:job/instance job))}]
+    (cond-> job-map
+            groups (assoc :groups (map #(str (:group/uuid %)) groups))
+            application (assoc :application {:name (:application/name application)
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/api.clj]: See [remove-datomic-namespacing](https://github.com/twosigma/Cook/blob/master/scheduler/src/cook/mesos/util.clj#L89) so you don't have to do this manually.

File: scheduler/src/cook/mesos/schema.clj [hunk 4]
```diff
@@ -117,6 +117,11 @@
     :db/valueType :db.type/ref
     :db/cardinality :db.cardinality/one
     :db.install/_attribute :db.part/db}
+   {:db/id (d/tempid :db.part/db)
+    :db/ident :job/application
+    :db/valueType :db.type/ref
+    :db/cardinality :db.cardinality/one
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/schema.clj]: set `:db/isComponent true` (see comment above)

File: scheduler/src/cook/mesos/schema.clj [hunk 5]
```diff
@@ -332,27 +337,40 @@
...
+    :db.install/_attribute :db.part/db}
+   {:db/id (d/tempid :db.part/db)
+    :db/ident :label/value
+    :db/valueType :db.type/string
+    :db/cardinality :db.cardinality/one
+    :db.install/_attribute :db.part/db}
+   ;; Application attributes
+   {:db/id (d/tempid :db.part/db)
+    :db/doc "TODO(DPO)"
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/schema.clj]: TODO

File: scheduler/src/cook/mesos/schema.clj [hunk 6]
```diff
@@ -332,27 +337,40 @@
...
+    :db/ident :application/name
+    :db/valueType :db.type/string
+    :db/cardinality :db.cardinality/one
+    :db.install/_attribute :db.part/db}
+   {:db/id (d/tempid :db.part/db)
+    :db/doc "TODO(DPO)"
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/schema.clj]: TODO

File: scheduler/src/cook/mesos/api.clj (hunk 1 above)
Comment (dposada) [inline scheduler/src/cook/mesos/api.clj]: Agreed, but it seems like we shouldn't break people who are sending empty string names

File: scheduler/src/cook/mesos/api.clj (hunk 2 above)
Comment (dposada) [inline scheduler/src/cook/mesos/api.clj]: Thanks, will take a look

File: scheduler/src/cook/mesos/api.clj (hunk 3 above)
Comment (dposada) [inline scheduler/src/cook/mesos/api.clj]: Nice! Will do

File: scheduler/src/cook/mesos/api.clj (hunk 2 above)
Comment (dposada) [inline scheduler/src/cook/mesos/api.clj]: Done, thanks

**Comment by dposada** [comment]
@wyegelwel I assume we want to update the `JobClient` to allow for specifying the application info as well?

**Comment by wyegelwel** [comment]
Yea it is probably worth doing. A lot of clients use it.

**Pull request merged**